src/mobiletwinned.cpp

```diff
@@ -140,17 +140,30 @@ void MobileTwinnedAppServerTsx::on_initial_request(pjsip_msg* req)
   // Set up the fork to the VoIP client.
   // To prevent both the VoIP client and native mobile service both
   // ringing on the same device, add a Reject-Contact header containing
-  // +sip.phone to the request headed for the VoIP client.
+  // +sip.with-twin to the request headed for the VoIP client. (We
+  // require that Gemini-compatible clients set this if they detect
+  // that they are colocated with a native client).
   LOG_DEBUG("Creating forked request to VoIP client");
   pj_pool_t* voip_pool = get_pool(voip_req);
   pjsip_reject_contact_hdr* reject_hdr =
                                    pjsip_reject_contact_hdr_create(voip_pool);
   pjsip_param* phone = PJ_POOL_ALLOC_T(voip_pool, pjsip_param);
-  phone->name = STR_PHONE;
+  phone->name = STR_WITH_TWIN;
   phone->value.slen = 0;
   pj_list_insert_after(&reject_hdr->feature_set, phone);
   pjsip_msg_add_hdr(voip_req, (pjsip_hdr*)reject_hdr);
 
+  // We also need a Reject-Contact header containg "g.3gpp.ics", to
+  // ensure that this never matches a native client without a
+  // colocated VoIP phone (which should be rung by the other fork). 
+  pjsip_reject_contact_hdr* reject_hdr2 =
+                                   pjsip_reject_contact_hdr_create(voip_pool);
+  pjsip_param* reject_native = PJ_POOL_ALLOC_T(voip_pool, pjsip_param);
+  reject_native->name = STR_3GPP_ICS;
+  reject_native->value.slen = 0;
+  pj_list_insert_after(&reject_hdr2->feature_set, reject_native);
+  pjsip_msg_add_hdr(voip_req, (pjsip_hdr*)reject_hdr2);
+
   // Set up the fork to the native device.
   // Append the twin prefix (if set) to the request URI, and add an
   // Accept-Contact header specifying g.3gpp.ics.
```

Comment:
This needs `explicit` and `required` set (otherwise this fork goes to other VoIP endpoints).